retained message matching bug

"a/#" can match retained message "b/a"

matchTopic() cannot pass this test: 
```
func TestTrieDB_GetMatchedMessages(t *testing.T) {
	a := assert.New(t)
	s := NewStore()
	msgs := []*gmqtt.Message{
		{
			Topic:   "a/b",
			Payload: []byte{1, 2, 3},
		},
		{
			Topic:   "b/a",
			Payload: []byte{1, 2, 3},
		},
	}
	var tt = []struct {
		TopicFilter string
		expected    map[string]*gmqtt.Message
	}{
		{
			TopicFilter: "a/#",
			expected: map[string]*gmqtt.Message{
				"a/b": {
					Payload: []byte{1, 2, 3},
				},
			},
		},
	}
	for _, v := range msgs {
		s.AddOrReplace(v)
	}
	for _, v := range tt {
		t.Run(v.TopicFilter, func(t *testing.T) {
			rs := s.GetMatchedMessages(v.TopicFilter)
			a.Equal(len(v.expected), len(rs))
			got := make(map[string]*gmqtt.Message)
			for _, v := range rs {
				got[v.Topic] = v
			}
			for k, v := range v.expected {
				a.Equal(v.Payload, got[k].Payload)
			}
		})
	}
}
```

```
=== RUN   TestTrieDB_GetMatchedMessages
=== CONT  TestTrieDB_GetMatchedMessages
    trie_db_test.go:86: 
        	Error Trace:	trie_db_test.go:86
        	Error:      	Not equal: 
        	            	expected: 1
        	            	actual  : 2
        	Test:       	TestTrieDB_GetMatchedMessages
--- FAIL: TestTrieDB_GetMatchedMessages (0.00s)
=== RUN   TestTrieDB_GetMatchedMessages/a/#
    --- PASS: TestTrieDB_GetMatchedMessages/a/# (0.00s)


Expected :1
Actual   :2
<Click to see difference>


FAIL
```

--- a/retained/interface.go
+++ b/retained/interface.go
@@ -24,7 +24,7 @@ type Store interface {
 	Remove(topicName string)
 	// GetMatchedMessages returns the retained messages that match the passed topic filter.
 	GetMatchedMessages(topicFilter string) []*gmqtt.Message
-	// iterate iterate all retained messages. The callback is called once for each message.
+	// Iterate iterate all retained messages. The callback is called once for each message.
 	// If callback return false, the iteration will be stopped.
 	// Notice:
 	// The results are not sorted in any way, no ordering of any kind is guaranteed.

--- a/retained/trie/retain_trie.go
+++ b/retained/trie/retain_trie.go
@@ -64,11 +64,7 @@ func (t *topicTrie) matchTopic(topicSlice []string, fn retained.IterateFn) {
 	endFlag := len(topicSlice) == 1
 	switch topicSlice[0] {
 	case "#":
-		if t.parent != nil {
-			t.parent.preOrderTraverse(fn)
-		} else {
-			t.preOrderTraverse(fn)
-		}
+		t.preOrderTraverse(fn)
 	case "+":
 		// 当前层的所有
 		for _, v := range t.children {
@@ -79,9 +75,7 @@ func (t *topicTrie) matchTopic(topicSlice []string, fn retained.IterateFn) {
 			} else {
 				v.matchTopic(topicSlice[1:], fn)
 			}
-
 		}
-
 	default:
 		if n := t.children[topicSlice[0]]; n != nil {
 			if endFlag {
@@ -150,7 +144,9 @@ func (t *topicTrie) preOrderTraverse(fn retained.IterateFn) bool {
 		}
 	}
 	for _, c := range t.children {
-		c.preOrderTraverse(fn)
+		if !c.preOrderTraverse(fn) {
+			return false
+		}
 	}
-	return false
+	return true
 }

--- a/retained/trie/trie_db_test.go
+++ b/retained/trie/trie_db_test.go
@@ -72,6 +72,10 @@ func TestTrieDB_GetMatchedMessages(t *testing.T) {
 			Payload: []byte{1, 2, 3},
 		},
 		{
+			Topic:   "b/a",
+			Payload: []byte{1, 2, 3},
+		},
+		{
 			Topic:   "a",
 			Payload: []byte{1, 2, 3},
 		},
@@ -80,6 +84,14 @@ func TestTrieDB_GetMatchedMessages(t *testing.T) {
 		TopicFilter string
 		expected    map[string]*gmqtt.Message
 	}{
+		{
+			TopicFilter: "a/+/c/",
+			expected: map[string]*gmqtt.Message{
+				"a/b/c/": {
+					Payload: []byte{1, 2, 3, 4},
+				},
+			},
+		},
 		{
 			TopicFilter: "a/+",
 			expected: map[string]*gmqtt.Message{
@@ -104,6 +116,9 @@ func TestTrieDB_GetMatchedMessages(t *testing.T) {
 					Payload: []byte{1, 2, 3},
 				},
 				"a/b": {
+					Payload: []byte{1, 2, 3},
+				},
+				"b/a": {
 					Payload: []byte{1, 2, 3},
 				},
 				"a": {
@@ -198,6 +213,10 @@ func TestTrieDB_Iterate(t *testing.T) {
 			Topic:   "a",
 			Payload: []byte{1, 2, 3},
 		},
+		{
+			Topic:   "$SYS/a/b",
+			Payload: []byte{1, 2, 3},
+		},
 	}
 
 	for _, v := range msgs {
@@ -209,5 +228,47 @@ func TestTrieDB_Iterate(t *testing.T) {
 		return true
 	})
 	a.ElementsMatch(msgs, rs)
+}
+
+func TestTrieDB_Iterate_Cancel(t *testing.T) {
+	a := assert.New(t)
+	s := NewStore()
+	msgs := []*gmqtt.Message{
+		{
+			Topic:   "a/b/c/d",
+			Payload: []byte{1, 2, 3},
+		},
+		{
+			Topic:   "a/b/c/",
+			Payload: []byte{1, 2, 3, 4},
+		},
+		{
+			Topic:   "a/",
+			Payload: []byte{1, 2, 3},
+		},
+		{
+			Topic:   "a/b",
+			Payload: []byte{1, 2, 3},
+		},
+		{
+			Topic:   "a",
+			Payload: []byte{1, 2, 3},
+		},
+	}
+
+	for _, v := range msgs {
+		s.AddOrReplace(v)
+	}
+	var i int
+	var rs []*gmqtt.Message
+	s.Iterate(func(message *gmqtt.Message) bool {
+		if i == 2 {
+			return false
+		}
+		rs = append(rs, message)
+		i++
+		return true
+	})
+	a.Len(rs, 2)
 
 }

